core/string_utils.cpp

```diff
@@ -60,6 +60,50 @@ UString jsonnet_string_escape(const UString &str, bool single)
     return ss.str();
 }
 
+unsigned long jsonnet_string_parse_unicode(const LocationRange &loc, const char32_t *c)
+{
+    unsigned long codepoint = 0;
+    // Expect 4 hex digits.
+    for (unsigned i = 0; i < 4; ++i) {
+        auto x = (unsigned char)(c[i]);
+        unsigned digit;
+        if (x == '\0') {
+            auto msg = "Truncated unicode escape sequence in string literal.";
+            throw StaticError(loc, msg);
+        } else if (x >= '0' && x <= '9') {
+            digit = x - '0';
+        } else if (x >= 'a' && x <= 'f') {
+            digit = x - 'a' + 10;
+        } else if (x >= 'A' && x <= 'F') {
+            digit = x - 'A' + 10;
+        } else {
+            std::stringstream ss;
+            ss << "Malformed unicode escape character, "
+               << "should be hex: '" << x << "'";
+            throw StaticError(loc, ss.str());
+        }
+        codepoint *= 16;
+        codepoint += digit;
+    }
+    return codepoint;
+}
+
+bool is_bmp_codepoint(const unsigned long codepoint)
+{
+    return codepoint != 0xd83d;
```

Comment:
This is insufficient. E.g. 0xd834 is also outside bmp. IIUC 0x0000 to 0xD7FF and 0xE000 to 0xFFFF are BMP characters and everything else is non-BMP.

Reply:
Yes, I was really just raising this as a hacky demo of the kinds of changes I thought might be needed. The way I split out the \u escape parsing is pretty vile.